File: Python/dijkstra/BlackBox/BlackBox.py
import heapq
import random

def dijkstra(graph, start):
    """
    Dijkstra's Algorithm: finds the shortest path from `start` to all other nodes in `graph`

    Args:
        graph: a dictionary of adjacency lists, where each key is a node and each value is a list of tuples (neighbor, weight)
        start: the starting node

    Returns:
        A dictionary of shortest distances from `start` to all other nodes, and a dictionary of predecessors
    """
    # Initialize distances and predecessors
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    predecessors = {node: None for node in graph}

    # Priority queue of nodes to visit, with their distances
    queue = [(0, start)]

    while queue:
        # Extract the node with the minimum distance from the queue
        current_distance, current_node = heapq.heappop(queue)

        # If we've already processed this node, skip it
        if current_distance > distances[current_node]:
            continue

        # Visit all neighbors of the current node
        for neighbor, weight in graph[current_node]:
            distance = current_distance + weight

            # If we've found a shorter path to the neighbor, update its distance and predecessor
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                predecessors[neighbor] = current_node
                heapq.heappush(queue, (distance, neighbor))

    return distances, predecessors

if __name__ == "__main__":
    # Generate a random graph with 1000 nodes
    num_nodes = 1000
    nodes = [f"Node_{i}" for i in range(num_nodes)]
    graph = {node: [] for node in nodes}

    # Add random edges with weights
    for node in nodes:
        num_edges = random.randint(1, 10)  # Each node has 1 to 10 neighbors
        neighbors = random.sample(nodes, num_edges)
        for neighbor in neighbors:
            if neighbor != node:
                graph[node].append((neighbor, random.randint(1, 100)))  # Random weight between 1 and 100

    # Select a random start node
    start_node = random.choice(nodes)
    # Run Dijkstra's algorithm
    dijkstra(graph, start_node)

File: C/dijkstra/BlackBox/BlackBox.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define V 1000 // Number of vertices in the graph

// Function to find the vertex with the minimum distance value
int minDistance(int dist[], int sptSet[]) {
    int min = INT_MAX, min_index;

    for (int v = 0; v < V; v++) {
        if (sptSet[v] == 0 && dist[v] <= min) {
            min = dist[v];
            min_index = v;
        }
    }
    return min_index;
}

// Function that implements Dijkstra's algorithm for a graph represented using an adjacency matrix
void dijkstra(int graph[V][V], int src) {
    int dist[V]; // Output array dist[i] holds the shortest distance from src to j
    int sptSet[V]; // sptSet[j] will be true if vertex j is included in the shortest path tree

    // Initialize all distances as INFINITE and sptSet[] as false
    for (int i = 0; i < V; i++) {
        dist[i] = INT_MAX;
        sptSet[i] = 0;
    }

    // Distance from source to itself is always 0
    dist[src] = 0;

    // Find the shortest path for all vertices
    for (int count = 0; count < V - 1; count++) {
        int u = minDistance(dist, sptSet);
        sptSet[u] = 1;

        // Update dist value of the adjacent vertices of the picked vertex
        for (int v = 0; v < V; v++) {
            if (!sptSet[v] && graph[u][v] && dist[u] != INT_MAX && dist[u] + graph[u][v] < dist[v]) {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }
}

int main() {
    // Create a graph with 1000 vertices
    int graph[V][V];

    // Initialize the graph with some values
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            if (i != j) {
                graph[i][j] = rand() % 10 + 1; // Random weights between 1 and 10
            } else {
                graph[i][j] = 0; // Distance to itself is 0
            }
        }
    }

    // Run Dijkstra's algorithm from the source vertex 0
    dijkstra(graph, 0);

    return 0;
}